src/stats/cdf.rs

```diff
@@ -0,0 +1,209 @@
+            *last_prob = cdf_prob;
+            Some((value, prob))
+        }
+        self.inner.iter().scan(f64::NEG_INFINITY, cdf_to_pmf)
+    }
+
+    /// Get cumulative probability for a given value. If the value is not present,
+    /// return the probability of the previous value. Complexity O(log n).
+    pub fn get(&self, value: &T) -> Option<LogProb> {
+        if self.inner.is_empty() {
+            None
+        }
+        else {
+            Some(match self.inner.binary_search_by(|e| e.0.partial_cmp(value).unwrap()) {
+                Ok(i) => self.inner[i].1,
+                Err(i) => if i > 0 { self.inner[i - 1].1 } else { f64::NEG_INFINITY }
+            })
+        }
+    }
+
+    /// Get probability (i.e. probability mass) for a given value. Complexity O(log n).
+    pub fn get_pmf(&self, value: &T) -> Option<LogProb> {
+        if self.inner.is_empty() {
+            None
+        }
+        else {
+            Some(match self.inner.binary_search_by(|e| e.0.partial_cmp(value).unwrap()) {
+                Ok(i) => if i > 0 { logprobs::sub(self.inner[i].1, self.inner[i - 1].1) } else { self.inner[0].1 },
+                Err(i) => if i > 0 { self.inner[i - 1].1 } else { f64::NEG_INFINITY }
+            })
+        }
+    }
+
+    /// Return total probability.
+    pub fn total_prob(&self) -> LogProb {
+        let &(_, prob) = self.inner.last().unwrap();
+        prob
+    }
+
+    /// Return maximum a posteriori probability estimate (MAP).
+    pub fn map(&self) -> &T {
+        let mut max = self.iter_pmf().next().unwrap();
+        for e in self.iter_pmf() {
+            if e.1 >= max.1 {
+                max = e;
+            }
+        }
+        &max.0
+    }
+
+    /// Return 95% credible interval.
+    pub fn credible_interval(&self) -> (&T, &T) {
+        let lower = self.inner.binary_search_by(|&(_, p)| p.partial_cmp(&0.025f64.ln()).unwrap()).unwrap_or_else(|i| i);
+        let upper = self.inner.binary_search_by(|&(_, p)| p.partial_cmp(&0.975f64.ln()).unwrap()).unwrap_or_else(|i| i - 1);
+
+        (&self.inner[lower].0, &self.inner[upper].0)
+    }
+
+    /// Number of entries in the CDF.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+}
+
+
+impl<T: NumCast + Clone + PartialOrd> CDF<T> {
+    pub fn expected_value(&self) -> f64 {
+        self.iter_pmf().map(|(value, prob)| {
+            cast::<T, f64>(value.clone()).unwrap() * prob.exp()
+        }).fold(0.0f64, |s, e| s + e)
+    }
+
+    pub fn variance(&self) -> f64 {
+        let ev = self.expected_value();
+        self.iter_pmf().map(|(value, prob)| {
+                (cast::<T, f64>(value.clone()).unwrap() - ev).powi(2) * prob.exp()
+        }).fold(0.0, |s, e| s + e)
+    }
+
+    pub fn standard_deviation(&self) -> f64 {
+        self.variance().sqrt()
+    }
+}
+
+pub type CDFPMFIter<'a, T> = iter::Scan<slice::Iter<'a, (T, LogProb)>, LogProb, fn(&mut LogProb, &'a (T, LogProb)) -> Option<(&'a T, LogProb)>>;
```

Comment:
As far as I can tell you only use this on line 97. I'd move it so that the definition was next to the usage. Also, this type is horrible-- I'm so excited for `impl Trait` 😄